src/fiber/channel.rs

```diff
@@ -1006,33 +1007,65 @@ where
+                                self.store.get_invoice_preimage(&payment_hash).is_some();
+                            if is_active && !is_settled {
+                                // This TLC is added to applied_add_tlcs in above, but
+                                // TLCs in the list applied_add_tlcs wouldn't be processed again.
+                                // For the unsettled active hold invoice TLCs, we should process them indefinitely
+                                // until they expire or are settled.
+                                state.tlc_state.applied_add_tlcs.remove(&add_tlc.tlc_id);
+                            }
+                            if status == CkbInvoiceStatus::Open {
+                                self.store
+                                    .update_invoice_status(
+                                        &payment_hash,
+                                        CkbInvoiceStatus::Received,
+                                    )
+                                    .expect("update invoice status failed");
+                            }
+                            if let Err(e) = self.store.add_invoice_channel_info(
+                                &payment_hash,
+                                InvoiceChannelInfo::new(state.get_id(), received_amount),
+                            ) {
+                                error!("Failed to add invoice channel mapping: {:?}", e);
+                            }
+                            // The updating of hold invoice is always done in settle_invoice rpc call.
+                            // So we can return early here.
+                            return Ok(());
+                        }
+                    }
                 }
+                Some(preimage) => preimage,
+            };
+
+            let filled_payment_hash: Hash256 = add_tlc.hash_algorithm.hash(preimage).into();
+            if add_tlc.payment_hash != filled_payment_hash {
+                return Err(ProcessingChannelError::FinalIncorrectPreimage);
+            }
+            // update invoice status to received only all the error checking passed
+            if let Some(_invoice) = self.store.get_invoice(&payment_hash) {
                 self.store
-                    .insert_payment_preimage(payment_hash, preimage)
-                    .map_err(|_| {
-                        ProcessingChannelError::InternalError("insert preimage failed".to_string())
-                    })?;
-            } else {
-                return Err(ProcessingChannelError::FinalIncorrectPaymentHash);
+                    .update_invoice_status(&payment_hash, CkbInvoiceStatus::Received)
+                    .expect("update invoice status failed");
             }
+            self.store
+                .insert_payment_preimage(payment_hash, preimage)
+                .map_err(|_| {
+                    ProcessingChannelError::InternalError("insert preimage failed".to_string())
+                })?;
```

Comment:
seems `ProcessingChannelError::FinalIncorrectPaymentHash` does not return to caller. we should return the error at line 1019?

Reply:
The store interface is a little bit confusing to me. Most of the store traits will never return errors. I don't know under what situation `insert_payment_preimage` may return an error (the current code does not return any error https://github.com/contrun/fiber/blob/fed2e63c7ec248d7319449d9bd7f388f3a943224/src/store/store.rs#L479-L488). It was my understanding that failure at this stage should be internal errors related to the store saving (not data validation error like the hashing result of this preimage does not match the hash). Can you recall why the `InvoiceStore` has `Result` as its return value? I thought all the business logic (data validation etc) should be done by the caller of the store trait (instead of in the store trait).

Reply:
I changed this to `FinalIncorrectPaymentHash` anyway.